In [54]:
import pandas as pd
import numpy as np
import os
import pickle
import time
import optuna
import warnings

from category_encoders import TargetEncoder
from xgboost import XGBRegressor
from scipy.stats import randint, uniform
from lightgbm import LGBMRegressor


from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import Lasso, LinearRegression
from sklearn.model_selection import train_test_split, cross_val_score, RandomizedSearchCV, _search_successive_halving, GridSearchCV
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor, ExtraTreesRegressor, HistGradientBoostingRegressor 
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error

warnings.filterwarnings("ignore")


In [55]:
cleanData = pd.read_csv('./data csv/kijkcijfersWeer.csv')
cleanData.sample(5)

,Unnamed: 0,FullDate,date,hour,Kanaal,Programma,Lengte_sec,Kijkers,Temperatuur,Gevoelstemp,Regen,Sneeuw,Weercode,Bewolking,Windsnelheid,Zonnenschijn,isFeestdag,Weekdag,isWeekend,Seizoen
37149,37149,2021-11-24 21:00:38,2021-11-24,21,VTM2,VB. C1 UEFA - CL.BRUGGE/LEIPZIG,5760,482798,4.3,1.9,0.0,0.0,3,100,6.4,0.0,0,2,0,herfst
19695,19695,2019-06-17 18:20:15,2019-06-17,18,VTM,DE BUURTPOLITIE,1582,203679,23.6,22.9,0.0,0.0,1,31,12.4,3600.0,0,0,0,lente
25490,25490,2020-04-04 21:34:12,2020-04-04,21,EEN,PRIDE & PREJUDICE,7231,290136,10.9,8.4,0.0,0.0,2,51,7.7,0.0,0,5,1,lente
31663,31663,2021-02-23 21:27:50,2021-02-23,21,EEN,WAAR IS MARK?,2040,675676,14.0,10.4,0.0,0.0,0,0,23.1,0.0,0,1,0,winter
3284,3284,2017-03-17 18:59:50,2017-03-17,18,VTM,NIEUWS 19U VTM,3092,606984,10.2,5.2,0.0,0.0,3,98,19.0,3600.0,0,4,0,winter


Feature engineering

In [56]:
def lagFeatures(df, n):
  for i in range(1,n+1):
    df[f'KijkersLag{i}'] = df.sort_values('FullDate').groupby('Programma')['Kijkers'].shift(i).fillna(df.groupby('Programma')['Kijkers'].transform('mean'))
  return df

In [57]:
cleanData = lagFeatures(cleanData, 3)
cleanData[cleanData['Programma'] == 'THUIS'][['FullDate', 'Kijkers', 'KijkersLag1', 'KijkersLag2', 'KijkersLag3']]


,FullDate,Kijkers,KijkersLag1,KijkersLag2,KijkersLag3
40,2016-10-03 20:14:23,1268561,1.059928e+06,1.059928e+06,1.059928e+06
60,2016-10-04 20:09:27,1169791,1.268561e+06,1.059928e+06,1.059928e+06
80,2016-10-05 20:13:52,1244502,1.169791e+06,1.268561e+06,1.059928e+06
121,2016-10-07 19:57:34,1156477,1.244502e+06,1.169791e+06,1.268561e+06
180,2016-10-10 19:59:00,1315826,1.156477e+06,1.244502e+06,1.169791e+06
...,...,...,...,...,...
61265,2025-04-04 20:24:13,779244,1.018053e+06,9.942260e+05,1.066599e+06
61325,2025-04-07 20:14:35,869193,7.792440e+05,1.018053e+06,9.942260e+05
61345,2025-04-08 20:13:09,880767,8.691930e+05,7.792440e+05,1.018053e+06
61366,2025-04-09 20:19:53,802132,8.807670e+05,8.691930e+05,7.792440e+05


Kardinaliteiten

In [58]:
#kardinaliteiten voor betere toepassingen
for kolom in cleanData.columns:
  print(f'{kolom}: {cleanData[kolom].nunique()}')

Unnamed: 0: 61405
FullDate: 61318
date: 3051
hour: 21
Kanaal: 29
Programma: 5963
Lengte_sec: 6774
Kijkers: 58860
Temperatuur: 414
Gevoelstemp: 478
Regen: 62
Sneeuw: 25
Weercode: 13
Bewolking: 101
Windsnelheid: 456
Zonnenschijn: 3555
isFeestdag: 2
Weekdag: 7
isWeekend: 2
Seizoen: 4
KijkersLag1: 58998
KijkersLag2: 56714
KijkersLag3: 55143


One Hot Encoding

In [59]:
#one hot encoding voor lage cardinaliteit
oneHotEnc = OneHotEncoder(handle_unknown="ignore")
lageKardinaliteit = cleanData[[ 'hour','Kanaal', 'isFeestdag', 'Weekdag', 'Seizoen']]
oneHot = oneHotEnc.fit_transform(lageKardinaliteit)
dfOneHot = pd.DataFrame(oneHot.toarray(), 
                            columns=oneHotEnc.get_feature_names_out(), 
                            index=lageKardinaliteit.index)

cleanData = cleanData.drop(columns=['hour', 'Kanaal', 'isFeestdag', 'Weekdag', 'Seizoen'])
cleanData = pd.concat([cleanData, dfOneHot], axis=1)

cleanData.sample(5)

,Unnamed: 0,FullDate,date,Programma,Lengte_sec,Kijkers,Temperatuur,Gevoelstemp,Regen,Sneeuw,...,Weekdag_1,Weekdag_2,Weekdag_3,Weekdag_4,Weekdag_5,Weekdag_6,Seizoen_herfst,Seizoen_lente,Seizoen_winter,Seizoen_zomer
24592,24592,2020-02-19 20:30:57,2020-02-19,DE AFSPRAAK,3220,288366,6.4,2.4,0.1,0.0,...,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0
51336,51338,2023-11-08 21:58:13,2023-11-08,"MOORDZAKEN, WETSDOKTERS",3287,379224,10.7,6.2,0.3,0.0,...,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0
17230,17230,2019-02-14 22:50:56,2019-02-14,HET JOURNAAL LAAT,931,340188,3.6,0.7,0.0,0.0,...,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0
18037,18037,2019-03-26 21:20:33,2019-03-26,TERUG NAAR RWANDA,2739,169798,5.1,2.4,0.0,0.0,...,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0
21630,21630,2019-09-24 22:52:47,2019-09-24,HET JOURNAAL LAAT,1139,356263,15.0,14.1,0.0,0.0,...,1.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0


In [60]:
#opslaan bij models
modelsDir = os.getcwd()
modelFile = os.path.join(modelsDir, ".",
                          "models", "oneHotEncoder.pkl")
os.makedirs(os.path.dirname(modelFile), exist_ok=True)
with open(modelFile, 'wb') as m:
  pickle.dump(oneHotEnc, m)

Target Encoding

In [61]:
#target encoding voor medium kardinaliteiten
targetEncoding = TargetEncoder()
medKardinaliteit = cleanData[['date', 'Programma', 'Lengte_sec', 'Temperatuur', 'Gevoelstemp', 'Regen', 'Bewolking', 'Windsnelheid', 'Zonnenschijn']]
#verdere feature engineering op vorig model
target = targetEncoding.fit_transform(medKardinaliteit, cleanData['Kijkers'])
target.sample(10)

,date,Programma,Lengte_sec,Temperatuur,Gevoelstemp,Regen,Bewolking,Windsnelheid,Zonnenschijn
53040,446625.82396,242060.656727,1657,9.2,5.1,0.0,100,22.4,0.00
33443,428042.67396,585853.225914,2803,14.4,9.6,0.0,100,23.7,2087.08
29832,501105.59896,317279.214567,6072,10.2,8.1,0.6,100,13.1,0.00
4947,448956.87396,273924.285714,1608,17.7,14.3,0.0,23,19.9,3600.00
30303,542070.39896,585853.225914,3316,8.6,5.3,0.0,44,16.1,0.00
12257,400990.27396,318450.257376,3429,20.6,20.1,0.0,81,15.9,3600.00
50059,435892.77396,233684.866438,1063,22.2,24.5,0.0,0,8.3,0.00
13860,414252.67396,697551.381773,1751,16.2,15.7,0.0,97,5.4,3397.82
19640,366118.17396,585853.225914,2518,18.1,18.4,0.0,43,3.2,3600.00
9955,498279.57396,427876.980663,5576,4.7,1.2,0.0,2,12.5,0.00


In [62]:
cleanData = cleanData.drop(columns=['date', 'Programma', 'Lengte_sec', 'Temperatuur', 'Gevoelstemp', 'Regen', 'Bewolking', 'Windsnelheid', 'Zonnenschijn'])
cleanData = pd.concat([cleanData, target], axis=1)
cleanData.sample(5)

,Unnamed: 0,FullDate,Kijkers,Sneeuw,Weercode,isWeekend,KijkersLag1,KijkersLag2,KijkersLag3,hour_0,...,Seizoen_zomer,date,Programma,Lengte_sec,Temperatuur,Gevoelstemp,Regen,Bewolking,Windsnelheid,Zonnenschijn
49509,49511,2023-08-09 19:41:31,600604,0.0,0,0,690291.000000,764542.000000,728610.000000,0.0,...,1.0,409610.92396,628487.089971,2195,18.9,17.2,0.0,0,9.9,3600.0
46492,46494,2023-03-10 20:06:04,518214,0.0,51,0,527890.000000,571927.000000,533312.000000,0.0,...,0.0,450105.59896,617095.220955,1515,2.9,-3.0,0.1,96,23.7,0.0
5073,5073,2017-06-15 23:09:13,192135,0.0,1,0,154203.333333,154203.333333,154203.333333,0.0,...,0.0,446149.44896,400821.814655,1759,17.3,14.5,0.0,38,19.6,0.0
15431,15431,2018-11-16 23:07:36,210736,0.0,0,0,191372.000000,376254.000000,250303.000000,0.0,...,0.0,445478.84896,289674.236832,973,4.8,2.3,0.0,0,7.7,0.0
43941,43942,2022-10-31 12:59:37,235901,0.0,3,0,195514.000000,173822.000000,181779.000000,0.0,...,0.0,445354.47396,242060.656727,1720,14.3,13.9,0.0,86,8.2,3600.0


Verbanden

In [63]:
dataOneHotTarget = cleanData.select_dtypes(include=[np.number])

verband = dataOneHotTarget.corr()

verband = verband['Kijkers'].abs().sort_values(ascending=False)

print(f'{verband.head(10)}')
dataOneHotTarget.shape

Kijkers        1.000000
KijkersLag1    0.894582
KijkersLag2    0.893195
KijkersLag3    0.875979
Programma      0.827215
hour_19        0.386184
Kanaal_EEN     0.333420
date           0.333361
hour_22        0.184308
hour_17        0.176407
Name: Kijkers, dtype: float64


(61405, 80)

In [64]:
# Eerst omzetten naar strings, daarna komma's verwijderen, en uiteindelijk naar floats
for col in ['KijkersLag1', 'KijkersLag2', 'KijkersLag3', 'Programma']:
    cleanData[col] = cleanData[col].astype(str).str.replace(',', '').astype(float)

# Print het resultaat
print(cleanData[['Kijkers', 'KijkersLag1', 'KijkersLag2', 'KijkersLag3', 'Programma']].head(10))


   Kijkers    KijkersLag1    KijkersLag2    KijkersLag3      Programma
0   721850  892001.072096  892001.072096  892001.072096  892001.072096
1   709606  628487.089971  628487.089971  628487.089971  628487.089971
2   548239  478421.179487  478421.179487  478421.179487  474186.638952
3   523610  791862.682708  791862.682708  791862.682708  791862.682708
4   496216  491224.740741  491224.740741  491224.740741  491217.906057
5   447427  351454.000000  351454.000000  351454.000000  431290.191170
6   424041  585853.225914  585853.225914  585853.225914  585853.225914
7   369066  408691.165611  408691.165611  408691.165611  408691.165611
8   368549  336230.500000  336230.500000  336230.500000  424185.509591
9   360544  332870.588235  332870.588235  332870.588235  332901.616764


In [65]:
#opslaan bij models
targetModelDir = os.getcwd()
targetModelPath = os.path.join(targetModelDir, ".",
                          "models", "oneHotTarget.pkl")
os.makedirs(os.path.dirname(targetModelPath), exist_ok=True)
with open(targetModelPath, 'wb') as m:
  pickle.dump(targetEncoding, m)

Model tests

In [53]:
def test_models(models, X, y, cv=10, scoring='neg_mean_absolute_error'):
    res = {}
    for m in models:
        model = type(m).__name__
        scores = cross_val_score(m, X, y, cv=cv, scoring=scoring)
        res[model] = {
            'scores': scores,
            'mean_score': scores.mean()
        }
        print(f"Scores {model}\n{scores}\nMAE = {scores.mean()}")
    return res

Lasso Regression

In [28]:
models = [
    RandomForestRegressor(verbose=True), Lasso()
]

# Zorg ervoor dat X en y goed geschaald zijn
scaler = StandardScaler()
X = dataOneHotTarget.drop(columns=['Kijkers', 'Unnamed: 0'])
X_scaled = scaler.fit_transform(X)
y = dataOneHotTarget['Kijkers']

# Run de test
results = test_models(models, X_scaled, y)

[Parallel(n_jobs=1)]: Done  49 tasks      | elapsed:  1.0min


KeyboardInterrupt: 

RandomForest heeft betere scores:
  
  RandomForest
  MAE = -52082.47210795064

  Lasso
  MAE = -64622.292775108515

Naar csv schrijven

In [ ]:
dir = os.getcwd()
pathFile = os.path.join(dir, "./data csv/oneHotTarget.csv")

os.makedirs(os.path.dirname(pathFile), exist_ok=True)

dataOneHotTarget.to_csv(pathFile)


keuze maken voor model

In [ ]:
#trainset en testset maken
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, shuffle=True)

def scoreModel(model,X,y, cv=5):
    print(f"Model: {model.__class__.__name__}")
    start = time.time()
        
    # MAE
    mae_scores = cross_val_score(model, X, y, cv=cv, scoring='neg_mean_absolute_error')
    mae = -np.mean(mae_scores)

    # MAPE
    mape_scores = cross_val_score(model, X, y, cv=cv, scoring='neg_mean_absolute_percentage_error')
    mape = -np.mean(mape_scores) * 100

    duration = time.time() - start
    min, sec = divmod(duration, 60)

    print(f"MAE:  {mae:.4f}")
    print(f"MAPE: {mape:.2f}%")
    print(f"Tijd: {int(min)}m {sec:.1f}s")

    return {'Model': model.__class__.__name__, 'MAE': mae, 'MAPE': mape, 'Tijd': duration}
def startModel(models, X, y, cv=5):
    results = []
    for model in models:
        result = scoreModel(model, X, y, cv=cv)
        results.append(result)
    return results

In [ ]:
linear = LinearRegression()
decision = DecisionTreeRegressor()
randomForest = RandomForestRegressor()
gradientBoosting = GradientBoostingRegressor()
xgb = XGBRegressor()
extraTrees = ExtraTreesRegressor()
histGB = HistGradientBoostingRegressor()

startModel([linear, decision, randomForest, gradientBoosting, xgb, extraTrees, histGB], X, y)


Model: LinearRegression
MAE:  65143.6876
MAPE: 20.22%
Tijd: 0m 1.5s
Model: DecisionTreeRegressor


KeyboardInterrupt: 

# Best models:

## Model: ExtraTreesRegressor

MAE:  51912.1192

MAPE: 15.50%

Tijd: 11m 29.4s

## Model: HistGradientBoostingRegressor

MAE:  52753.3521

MAPE: 15.64%

Tijd: 0m 11.3s

## Model: XGBRegressor

MAE:  53561.6539

MAPE: 15.62%

Tijd: 0m 4.2s

## RandomForest er niet bij want 18m+ traintijd voor slechts 200 MAE verschil met XGB

LGBM zou ook een goede optie moetten zijn


ExtraTrees: Beste nauwkeurigheid maar lange training tijd

HistGradientBoosting en XGB: Veel sneller met minimale performance drop

Verbeteren van de top 3 modellen

In [66]:
warnings.filterwarnings("ignore")

#parameters
model_param = {
    # 'XGBoost': {
    #     'model': XGBRegressor(),
    #     'params': {
    #         'n_estimators': randint(100, 1000),
    #         'learning_rate': uniform(0.01, 0.3),
    #         'max_depth': randint(3, 15),
    #         'min_child_weight': randint(1, 10),
    #         'subsample': uniform(0.6, 0.4),
    #         'colsample_bytree': uniform(0.6, 0.4),
    #         'gamma': uniform(0, 0.5),
    #         'reg_lambda': uniform(0.1, 10),
    #         'reg_alpha': uniform(0, 10)
    #     }
    # },
    # 'HistGradient': {
    #     'model': HistGradientBoostingRegressor(),
    #     'params': {
    #         'learning_rate': uniform(0.01, 0.3),
    #         'max_iter': randint(100, 1000),
    #         'max_depth': randint(3, 15),
    #         'l2_regularization': uniform(0.1, 10),
    #         'min_samples_leaf': randint(1, 10),
    #         'max_leaf_nodes': randint(20, 100)
    #     }
    # },
    "LightGBM": {
        "model": LGBMRegressor(),
        "params": {
            "num_leaves": [31, 50, 70],
            "learning_rate": [0.01, 0.05, 0.1, 0.15],
            "n_estimators": [100, 200, 500],
            "max_depth": [-1, 5, 10, 20],
            "min_child_samples": [10, 20, 30],
            "subsample": [0.6, 0.8, 1.0],
            "colsample_bytree": [0.6, 0.8, 1.0],
            "reg_alpha": [0, 0.1, 0.5],
            "reg_lambda": [0, 0.1, 0.5]
        }
    }
}

extra = ExtraTreesRegressor()
hist = HistGradientBoostingRegressor()
xgb = XGBRegressor()
lgbm = LGBMRegressor()


Random Search

In [ ]:
for name, config in model_param.items():
    print(f"Model: {name}")
    model = config['model']
    param = config['params']
    
    rsLoop = RandomizedSearchCV(
        estimator=model,
        param_distributions=param,
        n_iter=250,
        cv=5,
        scoring='neg_mean_absolute_error',
        n_jobs=-1,
        verbose=2
    )
    
    start = time.time()
    rsLoop.fit(X_train, y_train)
    duration = time.time() - start

    print(f"Tijd {name}: {duration:.2f}s")
    print(f"Beste param {name}: {rsLoop.best_params_}")
    print(f"Beste MAE: {rsLoop.best_score_:.4f}")

Model: XGBoost
Fitting 5 folds for each of 250 candidates, totalling 1250 fits


KeyboardInterrupt: 

In [ ]:
best_params = rsLoop.best_params_
tuned_model = rsLoop.best_estimator_

y_pred = tuned_model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
mape = mean_absolute_percentage_error(y_test, y_pred)
accuracy = 1 - mape

print(f'Mean absolute error: {mae}')
print(f'Mean absolute percentage error: {mape}')
print(f'Accuracy: {accuracy}')

Mean absolute error: 47585.56428634255
Mean absolute percentage error: 0.14202693123140966
Accuracy: 0.8579730687685904


In [ ]:

rsExtra = RandomizedSearchCV(
  estimator=extra,
  param_distributions=model_param['ExtraTrees']['params'],
  n_iter=150,
  cv=5,
  scoring='neg_mean_absolute_error',
  n_jobs=-1,
  verbose=2
)
    
start = time.time()
rsExtra.fit(X_train, y_train)
duration = time.time() - start

print(f"Tijd: {duration:.2f}s")
print(f"Beste param: {rsExtra.best_params_}")
print(f"Beste MAE: {rsExtra.best_score_:.4f}")

Fitting 5 folds for each of 100 candidates, totalling 500 fits
Tijd: 2857.67s
Beste param: {'max_depth': 9, 'max_features': np.float64(0.9124558702471834), 'min_samples_leaf': 2, 'min_samples_split': 4, 'n_estimators': 742}
Beste MAE: -54224.2000


### XGBoost
Tijd XGBoost: 1334.04s
Beste param XGBoost: {'colsample_bytree': np.float64(0.6550054872406584), 'gamma': np.float64(0.08367508771505777), 'learning_rate': np.float64(0.04567558428183022), 'max_depth': 8, 'min_child_weight': 2, 'n_estimators': 609, 'reg_alpha': np.float64(6.924415940372761), 'reg_lambda': np.float64(0.2620475254770348), 'subsample': np.float64(0.970829180420445)}
Beste MAE: -46309.9074
### HistGradient
Tijd HistGradient: 1145.61s
Beste param HistGradient: {'l2_regularization': np.float64(6.796320445820958), 'learning_rate': np.float64(0.02650163315092701), 'max_depth': 9, 'max_iter': 770, 'max_leaf_nodes': 96, 'min_samples_leaf': 4}
Beste MAE: -47550.1241
### ExtraTree
Tijd: 2857.67s
Beste param: {'max_depth': 9, 'max_features': np.float64(0.9124558702471834), 'min_samples_leaf': 2, 'min_samples_split': 4, 'n_estimators': 742}
Beste MAE: -54224.2000
### LightGBM
Tijd LightGBM: 448.79s
Beste param LightGBM: {'subsample': 1.0, 'reg_lambda': 0, 'reg_alpha': 0.5, 'num_leaves': 50, 'n_estimators': 500, 'min_child_samples': 10, 'max_depth': 10, 'learning_rate': 0.1, 'colsample_bytree': 0.8}
Beste MAE: -47344.7920

In [ ]:
def run_optuna(model_name):
    def objective(trial):
        # if model_name == 'ExtraTrees':
        #     model = ExtraTreesRegressor(
        #         n_estimators=trial.suggest_int('n_estimators', 100, 500),
        #         max_depth=trial.suggest_int('max_depth', 5, 30),
        #         min_samples_split=trial.suggest_int('min_samples_split', 2, 10),
        #         min_samples_leaf=trial.suggest_int('min_samples_leaf', 1, 10),
        #         random_state=42
        #     )
        # if model_name == 'HistGradient':
        #     model = HistGradientBoostingRegressor(
        #         learning_rate=trial.suggest_float('learning_rate', 0.01, 0.3),
        #         max_iter=trial.suggest_int('max_iter', 200, 500),
        #         max_depth=trial.suggest_int('max_depth', 3, 15),
        #         max_leaf_nodes=trial.suggest_int('max_leaf_nodes', 20, 100),
        #         random_state=42
        #     )
        # if model_name == 'XGBoost':
        #     model = XGBRegressor(
        #         n_estimators=trial.suggest_int('n_estimators', 200, 800),
        #         max_depth=trial.suggest_int('max_depth', 10, 15),
        #         learning_rate=trial.suggest_float('learning_rate', 0.01, 0.2, log=True),
        #         subsample=trial.suggest_float('subsample', 0.5, 1.0),
        #         colsample_bytree=trial.suggest_float('colsample_bytree', 0.5, 1.0),
        #         reg_alpha=trial.suggest_float('reg_alpha', 0.0, 1.0),
        #         reg_lambda=trial.suggest_float('reg_lambda', 0.0, 1.0),
        #         min_child_weight=trial.suggest_int('min_child_weight', 1, 10),
        #         gamma=trial.suggest_float('gamma', 0, 5),
        #         random_state=42,
        #         n_jobs=-1
        # )
        if model_name == 'LightGBM':
            model = LGBMRegressor(
                n_estimators=trial.suggest_int('n_estimators', 200, 800),
                max_depth=trial.suggest_int('max_depth', -1, 15),
                num_leaves=trial.suggest_int('num_leaves', 20, 350),
                learning_rate=trial.suggest_float('learning_rate', 0.01, 0.3, log=True),
                subsample=trial.suggest_float('subsample', 0.6, 1.0),
                colsample_bytree=trial.suggest_float('colsample_bytree', 0.6, 1.0),
                reg_alpha=trial.suggest_float('reg_alpha', 0.0, 1.0),
                reg_lambda=trial.suggest_float('reg_lambda', 0.0, 1.0),
                min_child_samples=trial.suggest_int('min_child_samples', 5, 30),
                random_state=42,
                n_jobs=-1,
                verbosity=-1
            )
        model.fit(X_train, y_train)
        trial.set_user_attr("model", model)
        pred = model.predict(X_test)
        return mean_absolute_error(y_test, pred)
    return objective

In [69]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, shuffle=True)

for name, config in model_param.items():
  print(f"Optuna {name}")
  optres = optuna.create_study(direction="minimize")
  optres.optimize(run_optuna(name), n_trials=500)

  start = time.time()
  duration = time.time() - start

  print(f"Tijd: {duration:.2f}s")
  print(f"Beste params {name}: ", optres.best_params)
  print(f"Beste MAE {name}: ", optres.best_value)

[I 2025-05-14 15:13:38,022] A new study created in memory with name: no-name-c1627124-3e85-4816-a02a-9c79f04ed5ca


Optuna LightGBM


[I 2025-05-14 15:13:42,795] Trial 0 finished with value: 46462.8167704101 and parameters: {'n_estimators': 398, 'max_depth': 0, 'num_leaves': 203, 'learning_rate': 0.03244517265075398, 'subsample': 0.9933888676453719, 'colsample_bytree': 0.9549076598560055, 'reg_alpha': 0.20281431559477792, 'reg_lambda': 0.04197211344371865, 'min_child_samples': 8}. Best is trial 0 with value: 46462.8167704101.
[I 2025-05-14 15:13:45,672] Trial 1 finished with value: 46908.45233684311 and parameters: {'n_estimators': 609, 'max_depth': 8, 'num_leaves': 174, 'learning_rate': 0.04654607232809103, 'subsample': 0.7492974928497977, 'colsample_bytree': 0.6794608308583151, 'reg_alpha': 0.39758018305888365, 'reg_lambda': 0.2968434062789498, 'min_child_samples': 26}. Best is trial 0 with value: 46462.8167704101.
[I 2025-05-14 15:13:48,166] Trial 2 finished with value: 48015.29177987524 and parameters: {'n_estimators': 234, 'max_depth': 10, 'num_leaves': 319, 'learning_rate': 0.03184236484473436, 'subsample': 0.7

Tijd: 0.00s
Beste params LightGBM:  {'n_estimators': 783, 'max_depth': 14, 'num_leaves': 332, 'learning_rate': 0.031065357902710616, 'subsample': 0.6982593215455819, 'colsample_bytree': 0.717273545134999, 'reg_alpha': 0.7803700200347652, 'reg_lambda': 0.6629991966878577, 'min_child_samples': 9}
Beste MAE LightGBM:  45693.15149266192


extratrees word verder genegeerd door de zeer lange trainingtijden en slechte finetuning resultaten, hierdoor heb ik dit vervamgen met LGBM doordat dit gelijkaardig zou moetten testen als XGB op mijn model

Optuna XGBoost

Beste params XGBoost:  {'n_estimators': 800, 'max_depth': 11, 'learning_rate': 0.024791795598051402, 'subsample': 0.9152182561342729, 'colsample_bytree': 0.628872599960794, 'reg_alpha': 0.4587526558642877, 'reg_lambda': 0.9054148915932084, 'min_child_weight': 9, 'gamma': 0.03130859570095774}
[I 2025-04-16 18:45:13,612] A new study created in memory with name: no-name-047d6f2c-8d88-466b-bb37-959ef292a0b9
Beste MAE XGBoost:  45756.13671875

Optuna HistGradient

Beste params HistGradient:  {'learning_rate': 0.09360779076213302, 'max_iter': 403, 'max_depth': 12, 'max_leaf_nodes': 83}
Beste MAE HistGradient:  47610.78728722957

Optuna ExtraTrees

Beste params ExtraTrees:  {'n_estimators': 351, 'max_depth': 27, 'min_samples_split': 8, 'min_samples_leaf': 1}
Beste MAE ExtraTrees:  48229.70593565771

Optuna LightGBM

Beste params LightGBM:  {'n_estimators': 791, 'max_depth': 14, 'num_leaves': 334, 'learning_rate': 0.029456000685140836, 'subsample': 0.8303507682774592, 'colsample_bytree': 0.6505159652741844, 'reg_alpha': 0.20430689348688258, 'reg_lambda': 0.3218993752340186, 'min_child_samples': 13}
Beste MAE LightGBM:  45536.17331152257

In [77]:
print(optres.best_trial.user_attrs)


{'model': LGBMRegressor(colsample_bytree=0.717273545134999,
              learning_rate=0.031065357902710616, max_depth=14,
              min_child_samples=9, n_estimators=783, n_jobs=-1, num_leaves=332,
              random_state=42, reg_alpha=0.7803700200347652,
              reg_lambda=0.6629991966878577, subsample=0.6982593215455819,
              verbosity=-1)}


In [76]:
best_params = optres.best_params
tuned_model = optres.best_trial.user_attrs['model']

y_pred = tuned_model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
mape = mean_absolute_percentage_error(y_test, y_pred)
accuracy = 1 - mape

print(f'Mean absolute error: {mae}')
print(f'Mean absolute percentage error: {mape}')
print(f'Accuracy: {accuracy}')

Mean absolute error: 45693.15149266192
Mean absolute percentage error: 0.13544911193090078
Accuracy: 0.8645508880690992


## Best so far
### LightGBM
Mean absolute error: 45337.28982557999
Mean absolute percentage error: 0.1334235812403486
Accuracy: 0.8665764187596514

GridsearchCV

In [ ]:

#parameters
model_param = {
    'XGBoost': {
        'model': XGBRegressor(),
        'params': {
            'n_estimators': [850],
            'learning_rate': [0.025, 0.03],
            'max_depth': [8, 9],
            'min_child_weight': [1,2],
            'subsample': [1.0],
            'colsample_bytree': [0.8],
            'gamma': [0.035, 0.04],
            'reg_lambda': [3],
            'reg_alpha': range(0, 5, 1)
        }
    },
    'HistGradient': {
        'model': HistGradientBoostingRegressor(),
        'params': {
            'n_estimators': [850],
            'learning_rate': [0.01, 0.1],
            'max_iter': [100, 200],
            'max_depth': [3, 5],
            'l2_regularization': [0.1, 1],
            'min_samples_leaf': [1, 2],
            'max_leaf_nodes': [20, 40]
        }
    },
    'LightGBM': {
        'model': LGBMRegressor(),
        'params': {
            'n_estimators': [850],
            'num_leaves': [31, 50],
            'learning_rate': [0.01, 0.1],
            'n_estimators': [100, 200],
            'max_depth': [-1, 5],
            'min_child_samples': [10, 20],
            'subsample': [0.6, 0.8],
            'colsample_bytree': [0.6, 0.8],
            'reg_alpha': [0, 0.1],
            'reg_lambda': [0, 0.1]
        }
    }
}

hist = HistGradientBoostingRegressor()
xgb = XGBRegressor()
lgbm = LGBMRegressor()


In [ ]:
for name, config in model_param.items():
    print(f"Model: {name}")
    model = config['model']
    param = config['params']
    
    gs = GridSearchCV(
        estimator=model,
        param_grid=param,
        cv=5,
        scoring='neg_mean_absolute_error',
        n_jobs=-1,
        verbose=2
    )
    
    start = time.time()
    gs.fit(X_train, y_train)
    duration = time.time() - start

    print(f"Tijd {name}: {duration:.2f}s")
    print(f"Beste param {name}: {gs.best_params_}")
    print(f"Beste MAE: {gs.best_score_:.4f}")

Model: XGBoost
Fitting 5 folds for each of 80 candidates, totalling 400 fits


KeyboardInterrupt: 

Resultaten tuned model

In [75]:
best_params = gs.best_params_
tuned_model = gs.best_estimator_

y_pred = tuned_model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
mape = mean_absolute_percentage_error(y_test, y_pred)
accuracy = 1 - mape

print(f'Mean absolute error: {mae}')
print(f'Mean absolute percentage error: {mape}')
print(f'Accuracy: {accuracy}')

NameError: name 'gs' is not defined

### model opslaan met beste parameters

In [73]:
besteModel = LGBMRegressor(
    n_estimators=788,
    max_depth=0,
    num_leaves=213,
    learning_rate=0.027352468420388434,
    subsample=0.8257111258276627,
    colsample_bytree=0.7709277343834496,
    reg_alpha=0.7512362258288539,
    reg_lambda=0.23466677100487698,
    min_child_samples=8,
    random_state=42,
    n_jobs=-1,
    verbosity=-1
)

besteModel.fit(X_train, y_train)
pred = besteModel.predict(X_test)
mae = mean_absolute_error(y_test, pred)
print(f"MAE: {mae:.4f}")

MAE: 46021.7181


In [78]:
#opslaan bij models
lgbmModelDir = os.getcwd()
lgbmModelPath = os.path.join(lgbmModelDir, ".",
                          "models", "lightGBM.pkl")
os.makedirs(os.path.dirname(lgbmModelPath), exist_ok=True)
with open(lgbmModelPath, 'wb') as m:
    pickle.dump(besteModel, m)